### PR TITLE
adjust MetricsClientSendFailingSRE to fire if error budget is burned fast

### PR DESCRIPTION
**Pull request by RiRa12621**
making MetricsClientSendFailingSRE more fault tolerant
This assumes an SLO of telemeter itself of 90% and will only fire if the client is burning the corresponding error budget fast.
After an hour 2% of a months error budget would have been consumed.
Since there is not a whole lot to do on the SRE end, I am tempted to additionally reduce it to warning.
This only fires if the error rate has been high for the past 1h AND 5m allowing us to recover quickly.

### CODE REVIEW

**Comment by RiRa12621** [comment]
/cc @metalmatze 
can you sanity check this please?

**Comment by RiRa12621** [comment]
/assign @rogbas
obeying the bots

**Comment by metalmatze** [comment]
Nice!

/lgtm

**Comment by metalmatze** [comment]
Btw this is measuring exactly the official SLO of 90% availability in 30days for Telemeter. :+1:

**Comment by RiRa12621** [comment]
/retitle adjust MetricsClientSendFailingSRE to fire if error budget is burned fast

**Comment by RiRa12621** [comment]
this fixes #397

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (lisa) [inline deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:15]: @RiRa12621 Can you leave a comment explaining the formula and the numbers, especially where 14.4 is coming from?

Comment (metalmatze) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:15]: The 14.40 is from burn rate 2 within 1h and then up to the 30 day availability window. Thus `(2 * 24 * 28) / 100`.
Read more here: https://landing.google.com/sre/workbook/chapters/alerting-on-slos/

I also recently recorded a meeting in the SLO workgroup meeting where I explain most of this chapter. Let me know on Slack if you need the link.

Comment (rogbas) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:15]: @metalmatze thanks for the clarification

**Comment by rogbas** [comment]
@RiRa12621 it LGTM, as long as we have a comment explaining the rationale (or maybe linking the sre-book chapter)

**Comment by RiRa12621** [comment]
/hold
This is not urgent, and I can wait until after the session so everyone feels comfortable and has a recording to go back to in case it's unclear to someone in the future

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (lisa) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:15]: Please put that in the YAML file, too.

**Comment by RiRa12621** [comment]
/unhold
ready for review

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (metalmatze) [inline deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml]: Maybe make it a multi-line string:
```suggestion
    100 * 
        sum by (job) (rate(metricsclient_request_send{status_code="5xx"}[5m]))
    / 
        sum by (job) (rate(metricsclient_request_send[5m])) 
    > 14.4 * 0.1 
and
    100 * 
        sum by (job) (rate(metricsclient_request_send{status_code="5xx"}[1h]))
    /
        sum by (job) (rate(metricsclient_request_send[1h])) 
    > 14.4 * 0.1
```

This would roughly be the official style guidelines.

Comment (metalmatze) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml]: To make it multi-line YAML start the first line with `expr: |-`

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (cblecker) [inline deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:14]: Can we still get a brief, human readable explanation in a yaml comment (like https://github.com/openshift/managed-cluster-config/blob/20d225db7bb8dcc79e12dc9a2490b84478c844d3/deploy/sre-prometheus/100-managed-velero-operator.PrometheusRule.yaml#L18-L20), and then suggesting that for a more detailed explanation, see the readme?

Comment (RiRa12621) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:14]: We could. I'm not sure how to condense that so that it's meaningful but short enough for a comment.

Comment (RiRa12621) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:14]: any idea/recommendation @cblecker ?

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (georgettica) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml]: or we can do `(2 * 24 * 30) / 100` in the yaml

**Comment by RiRa12621** [comment]
/unhold
adjusted to make everything !200 an error. that is fine in our scenario since we're the sender, so even 4xx is our bad

File: deploy/sre-prometheus/README.md
Comment (dofinn) [inline deploy/sre-prometheus/README.md:88]:
```suggestion
`rate(metricsclient_request_send{status_code!="200"}[5m]) / rate(metricsclient_request_send[5m])`
```

File: deploy/sre-prometheus/README.md
Comment (dofinn) [inline deploy/sre-prometheus/README.md:95]:
```suggestion
`14.4 * 0.1`
```

File: deploy/sre-prometheus/README.md
Comment (dofinn) [inline deploy/sre-prometheus/README.md:99]:
```suggestion
`rate(metricsclient_request_send{status_code!="200"}[5m]) / rate(metricsclient_request_send[5m]) > 14.4 * 0,1` 
```

File: deploy/sre-prometheus/README.md
Comment (dofinn) [inline deploy/sre-prometheus/README.md:101]:
```suggestion
Now add the conditions together to cover both the short and long windows that allow for fast recovery and fast alerting
```

File: deploy/sre-prometheus/README.md
Comment (dofinn) [inline deploy/sre-prometheus/README.md:106]:
```suggestion
rate(metricsclient_request_send{status_code!="200"}[5m]) / rate(metricsclient_request_send[5m]) > 14.4 * 0.1
and
rate(metricsclient_request_send{status_code!="200"}[1h]) / rate(metricsclient_request_send[1h]) > 14.4 * 0.1
```

File: deploy/sre-prometheus/README.md
Comment (dofinn) [inline deploy/sre-prometheus/README.md:114]:
```suggestion
100 * rate(metricsclient_request_send{status_code!="200"}[5m]) / rate(metricsclient_request_send[5m]) > 14.4 * 0.1
and
100 * rate(metricsclient_request_send{status_code!="200"}[1h]) / rate(metricsclient_request_send[1h]) > 14.4 * 0.1
```

**Review by dofinn — COMMENTED** [review]
Sorry about all the corrections, the `,` were irritating me :)

I want to really walk through this query one more time before lgtm'ing.

File: deploy/sre-prometheus/README.md
Comment (RiRa12621) [reply deploy/sre-prometheus/README.md:18]: Yes but this please a bigger role when recovering from the alert. Because it takes significantly longer to reduce your error rate when looking at a 60m period as opposed to a 5m period

Comment (dofinn) [reply deploy/sre-prometheus/README.md:18]: I get that aspect. Now im confused about if this comment is explaining the initial time to alert or time to recover?

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (dofinn) [inline deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:14]:
```suggestion
      expr: sum by (job) (rate(metricsclient_request_send{status_code!="200"}[5m])) / sum by (job) (rate(metricsclient_request_send[5m])) > (6 - (1-0.9000) and sum by (job) (rate(metricsclient_request_send{status_code!="200"}[1h])) / sum by (job) (rate(metricsclient_request_send[1h])) > (6 - (1-0.90000)
```

Upon really digging into this promQL, I found the following:

- telemeter client only sends metrics every 5 minutes. This was validated using `sum(metricsclient_request_send) - sum(metricsclient_request_send offset 10m)` returning a `value=2`. 
- Using a 14.4 burn rate on an SLO of 90% will never trigger as it requires 2% of the error budget in a single hour to be consumed. Because a 100% outage consumes only 1.4% of the budget in that hour, this alert could never fire.
- Updated the burn rate to use 6 which will actually trigger and allow reset to occur as a binary using the 5m windows
- Updated the pass the SLO measurement rather then the error budget itself. 

With this, the 5m window becomes a binary switch allowing fast reset and the 1h window allows for up to 60% (7+ requests) of requests to fail for the hour before triggering.

File: deploy/sre-prometheus/README.md
Comment (RiRa12621) [reply deploy/sre-prometheus/README.md:18]: "We now combine that with the time to recover"
Both :-)

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (RiRa12621) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:14]: Worth to note that this is based on the 90% SLO of telemeter itself

**Comment by RiRa12621** [comment]
i think we're good to go now @dofinn

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (dofinn) [inline deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:13]:
```suggestion
    - alert: MetricsClientSendFailingSRE
    # metricsclient sends 1 request/5m, 12/60m, 288/day, 8064/28days.
    # SLO=90%, Error Budget=10% or 804 failed requests.
    # Send an alert when 2% of Error budget (16.2 requests) is burned within a 2 hour window. Why 2 hours? Because the combination of request frequency and SLO @90% requires atleast 2 hours to breach a 2% burn. 
    # The short window of 10 minutes allows for quick recovery once the issue is resolved. 
```

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (dofinn) [inline deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:15]:
```suggestion
      expr: sum by (job) (rate(metricsclient_request_send{status_code!="200"}[10m])) / sum by (job) (rate(metricsclient_request_send[10m])) > (16.12 - (1-0.90000)) and sum by (job) (rate(metricsclient_request_send{status_code!="200"}[2h])) / sum by (job) (rate(metricsclient_request_send[2h])) > (16.12 - (1-0.90000))
      for: 10m
```

**Review by dofinn — CHANGES_REQUESTED** [review]
Final review on this :)
Please run `make` if you agree.

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (dofinn) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:13]: cc @cblecker

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (RiRa12621) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:15]: How do you get to a 16 burn rate?

Comment (dofinn) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:15]: 2% of 806  is 16.12

File: deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml
Comment (cblecker) [reply deploy/sre-prometheus/100-telemeter-client.PrometheusRule.yaml:13]: lgtm

**Pull request merged**
